builtin/logical/pki/backend_test.go

```diff
@@ -139,13 +310,234 @@ func checkCertsAndPrivateKey(keyType string, usage certUsage, validity time.Dura
+				"max_path_length": 0,
+			},
+		},
+
+		logicaltest.TestStep{
+			Operation: logical.WriteOperation,
+			Path:      "config/ca/sign",
+			Data: map[string]interface{}{
+				"use_csr_values": true,
+				"csr":            string(csrPem),
+				"format":         "der",
+			},
+			ErrorOk: true,
+		},
+
+		logicaltest.TestStep{
+			Operation: logical.WriteOperation,
+			Path:      "config/ca/generate/root/exported",
+			Data: map[string]interface{}{
+				"common_name":     "Root Cert",
+				"ttl":             "180h",
+				"max_path_length": 1,
+			},
+		},
+		logicaltest.TestStep{
+			Operation: logical.WriteOperation,
+			Path:      "config/ca/sign",
+			Data: map[string]interface{}{
+				"use_csr_values": true,
+				"csr":            string(csrPem),
+				"format":         "der",
+			},
+			Check: func(resp *logical.Response) error {
+				certString := resp.Data["certificate"].(string)
+				if certString == "" {
+					return fmt.Errorf("no certificate returned")
+				}
+				certBytes, _ := base64.StdEncoding.DecodeString(certString)
+				certs, err := x509.ParseCertificates(certBytes)
+				if err != nil {
+					return fmt.Errorf("returned cert cannot be parsed: %v", err)
+				}
+				if len(certs) != 1 {
+					return fmt.Errorf("unexpected returned length of certificates: %d", len(certs))
+				}
+				cert := certs[0]
+
+				if cert.MaxPathLen != 0 {
+					return fmt.Errorf("max path length of %d does not match the requested of 3", cert.MaxPathLen)
+				}
+				if !cert.MaxPathLenZero {
+					return fmt.Errorf("max path length zero is not set")
+				}
+
+				// We need to set these as they are filled in with unparsed values in the final cert
+				csrTemplate.Subject.Names = cert.Subject.Names
+				csrTemplate.Subject.ExtraNames = cert.Subject.ExtraNames
+
+				switch {
+				case !reflect.DeepEqual(cert.Subject, csrTemplate.Subject):
+					return fmt.Errorf("cert subject\n%#v\ndoes not match csr subject\n%#v\n", cert.Subject, csrTemplate.Subject)
+				case !reflect.DeepEqual(cert.DNSNames, csrTemplate.DNSNames):
+					return fmt.Errorf("cert dns names\n%#v\ndoes not match csr dns names\n%#v\n", cert.DNSNames, csrTemplate.DNSNames)
+				case !reflect.DeepEqual(cert.EmailAddresses, csrTemplate.EmailAddresses):
+					return fmt.Errorf("cert email addresses\n%#v\ndoes not match csr email addresses\n%#v\n", cert.EmailAddresses, csrTemplate.EmailAddresses)
+				case !reflect.DeepEqual(cert.IPAddresses, csrTemplate.IPAddresses):
+					return fmt.Errorf("cert ip addresses\n%#v\ndoes not match csr ip addresses\n%#v\n", cert.IPAddresses, csrTemplate.IPAddresses)
+				}
+				return nil
+			},
+		},
+	}
+	return ret
+}
+
 // Generates steps to test out CA configuration -- certificates + CRL expiry,
 // and ensure that the certificates are readable after storing them
-func generateCASteps(t *testing.T) []logicaltest.TestStep {
+func generateCATestingSteps(t *testing.T, caCert, caKey string, intdata, reqdata map[string]interface{}) []logicaltest.TestStep {
```

Comment:
We should probably break this up into multiple tests. This whole file is getting a bit huge, and should probably be split into multiple test files